CompilerSource/parsing/parser.cpp

```diff
@@ -1950,7 +1950,29 @@ std::unique_ptr<AST::Node> TryParseControlExpression(SyntaxMode mode_) {
   }
 }
 
+std::unique_ptr<AST::Node> MaybeDeclaration() {
```

Comment:
I worry this doesn't give us the ability to indicate that parsing went wrong on our return. It seems that all branches either do not consume tokens or end in another TryParse, so I think this is okay, except if `TryParseEitherFunctionalCastOrDeclaration` were to return `nullptr` without consuming anything, we'd have accidentally swallowed the `::` token this routine was called with. Does that make sense? If `TryParseEitherFunctionalCastOrDeclaration` will necessarily return non-null when the token is `::`, we're fine.

In that case, I think this should be renamed to TryParseSomething or another, but then, it seems that what we handle here is just functional casts, declarations, and global new/delete, which might just be worth adding to `TryParseEitherFunctionalCastOrDeclaration` rather than creating a new method. That function's name is already Jupiterian, so I might recommend just `TryParseDeclarationOrTypeExpression`.

What do you think? I'm sure Dhruv has stronger opinions than either of us. :slightly_smiling_face:

Reply:
Agreed, didn't think of this case.